src/evm_parse.c

```diff
@@ -549,37 +549,44 @@ enum parse_rv parse_abi_call_data(struct EVM_ABI_state *const state,
       }
     }
 
-    state->state++;
-    initFixed(&state->argument_state.fixedState, sizeof(state->argument_state));
-
     if(meta->known_endpoint) {
+      state->state = ABISTATE_ARGUMENTS;
+      initFixed(&state->argument_state.fixedState, sizeof(state->argument_state));
       if(hasValue) REJECT("No currently supported methods are marked as 'payable'");
       char *method_name = PIC(meta->known_endpoint->method_name);
       ADD_PROMPT("Contract Call", method_name, strlen(method_name), strcpy_prompt);
       return PARSE_RV_PROMPT;
     } else {
-      state->state = ABISTATE_DONE;
-      // Probably we have to allow this, as the metamask constraint means _this_ endpoint will be getting stuff it doesn't understand a lot.
-      static char const isPresentLabel[]="Is Present (unsafe)";
-      set_next_batch_size(&meta->prompt, 2);
+      state->state = ABISTATE_UNRECOGNIZED;
       ADD_ACCUM_PROMPT("Transfer", output_evm_prompt_to_string);
-      ADD_PROMPT("Contract Data", isPresentLabel, sizeof(isPresentLabel), strcpy_prompt);
       return PARSE_RV_PROMPT;
     }
   }
 
   case ABISTATE_ARGUMENTS: {
-    if(state->argument_index >= meta->known_endpoint->parameters_count)
-      return PARSE_RV_DONE;
-    sub_rv = parseFixed(&state->argument_state.fixedState, input, ETHEREUM_WORD_SIZE); // TODO: non-word size values
+    if(state->state == ABISTATE_ARGUMENTS) {
```

Comment:
@alexfmpe when wouldn't this be true? isn't `state->state` what we are switching on?

Reply:
Fall-through. We also end up here when going `ABISTATE_SELECTOR -> ABISTATE_UNRECOGNIZED`.
Maybe this would be less confusing with `goto`s.

Reply:
Oh right now we're returning in both branches in `ABISTATE_SELECTOR`. No fall-through anymore